### PR TITLE
Options Refactoring

### PR DESCRIPTION
**Pull request by chaosmail**
Hi Sébastien,

I am still very active in finding a perfect structure for the Options, so I conitnued refactoring everything. This time I want your opinion before I write all the tests.

Best,
Christoph

### CODE REVIEW

File: src/factories/Axis.ts [hunk 1]
```diff
@@ -6,15 +6,11 @@ module n3Charts.Factory {
...
-    public static SIDE = {
-      X: 'x',
-      Y: 'y'
-    };
 
     constructor(public side: string) {
       super();
 
-      if (d3.values(Axis.SIDE).indexOf(side) === -1) {
+      if (!Utils.AxisOptions.isValidSide(side)) {
```
Comment (lorem--ipsum) [inline src/factories/Axis.ts]: :+1:

**Comment by lorem--ipsum** [comment]
I think the possible series's types should be attached as static properties to the Series class. The SeriesOptions shouldn't contain this since they're one way to create Series, but not the only one (I'm thinking from a dev point of view, not necessarily from a framework user point of view)

File: src/utils/Options.ts [hunk 2]
```diff
@@ -1,10 +1,10 @@
 module n3Charts.Utils {
   'use strict';
 
-  export interface IAxes {
-    x: AxisOptions;
-    y: AxisOptions;
-    y2?: AxisOptions;
+  export interface IAxesSet {
```
Comment (lorem--ipsum) [inline src/utils/Options.ts]: No need for the word "set" here, in my opinion : "axes" is the plural form of "axis" anyway.

**Comment by chaosmail** [comment]
yeah I thought the same, but there is no Series Class :smile:

File: src/utils/Options.ts [hunk 3]
```diff
@@ -30,72 +36,38 @@ module n3Charts.Utils {
...
-    static SERIES_TYPES = {
-      DOT: 'dot',
-      LINE: 'line',
-      AREA: 'area',
-      COLUMN: 'column'
-    };
-
     public series: SeriesOptions[];
-    public axes: IAxes;
+    public axes: IAxesSet;
```
Comment (lorem--ipsum) [inline src/utils/Options.ts]: I'd suggest either `axes: IAxes` or `axesSet: IAxesSet` for consistency's sake.

File: src/utils/Options.ts (hunk 2 above)
Comment (chaosmail) [inline src/utils/Options.ts]: I thought very long about this; for me it was confusing that it is not just axes (like multiple axis) but it has a special shape. It's an object in fact, that deserves a name

Comment (chaosmail) [inline src/utils/Options.ts]: that's what I thought

Comment (lorem--ipsum) [inline src/utils/Options.ts]: That's true. Let's keep this then.

**Comment by lorem--ipsum** [comment]
Oh there isn't ? Dammit, I'm already lost in the codebase ^^

**Comment by lorem--ipsum** [comment]
So I can merge this, do you want to do something about the axes: IAxes or axesSet: IAxesSet thing ?

**Comment by chaosmail** [comment]
I will write tests before we merge this ;)

**Comment by chaosmail** [comment]
Alright, here are some tests. Now this looks a little bit better for me.

**Comment by chaosmail** [comment]
You wait, I add more tests and slightly more refactoring. You should really check it out - I mostly worked on the Options.ts and SeriesOptions.ts files. I am already quite happy with the result :)

File: src/utils/Options.ts [hunk 4]
```diff
@@ -104,56 +76,74 @@ module n3Charts.Utils {
...
+    getSeriesByType(type: string) {
+      if (!SeriesOptions.isValidType(type)) {
         throw new TypeError('Unknown series type: ' + type);
       }
 
-      return this.series.filter((series) => series.type.indexOf(type) !== -1);
+      return this.series.filter((s) => s.hasType(type));
+    }
+
+    static getBoolean(val: any, def: boolean = true) {
```
Comment (lorem--ipsum) [inline src/utils/Options.ts]: I'm not fan of abbreviating variables, especially in methods' signature... Do you mind changing this ?

File: src/utils/SeriesOptions.ts [hunk 5]
```diff
@@ -9,64 +20,117 @@ module n3Charts.Utils {
...
     getMainType() {
-      if (this.type.length === 1) {
-        return this.type[0];
-      }
-
-      var types = Options.SERIES_TYPES;
+      // @Sebastien: do we really need a main type?
+      // I saw we are just using it for defining classes,
+      // but this could be perfectly done for every type
+      d3.values(SeriesOptions.TYPE).forEach((type) => {
```
Comment (lorem--ipsum) [inline src/utils/SeriesOptions.ts]: I think the `getMainType()` is used to set the series' svg's class name...

File: src/utils/SeriesOptions.ts [hunk 6]
```diff
@@ -9,64 +20,117 @@ module n3Charts.Utils {
...
-      if (this.type.indexOf(types.AREA) !== -1) {
-        return types.AREA;
-      }
-
-      if (this.type.indexOf(types.LINE) !== -1) {
-        return types.LINE;
-      }
+      // We should always find the type with the lines
+      // above and never reach this return statement
+      return this.type[0];
```
Comment (lorem--ipsum) [inline src/utils/SeriesOptions.ts]: Then let's throw an error !

File: src/utils/SeriesOptions.ts [hunk 7]
```diff
@@ -9,64 +20,117 @@ module n3Charts.Utils {
...
+    }
 
-      if (this.type.indexOf(types.DOT) !== -1) {
-        return types.DOT;
-      }
+    /**
+     * Returns true if the series has a type *type*,
+     * where type should be a value of SeriesOptions.TYPE
```
Comment (lorem--ipsum) [inline src/utils/SeriesOptions.ts]: If `type` should be a value of `SeriesOptions.TYPE`, let's remove the comment and enforce the rule by throwing an error if it's not the case !

File: test/spec/utils/Options.spec.ts [hunk 8]
```diff
@@ -18,181 +18,168 @@ describe('n3Charts.Utils.Options', () => {
...
-
-      options.parse(arg);
+    it('should create a margin property with the type object', () => {
+      options = new Options();
 
-      var testing = spy.calledWith(Options.DEFAULT.series);
+      var testing = angular.isObject(options.margin);
       var expected = true;
 
       expect(testing).to.equal(expected);
```
Comment (lorem--ipsum) [inline test/spec/utils/Options.spec.ts]: Maybe `expect(options.margin).to.be.a(Object)` would be more explicit ?

File: src/utils/Options.ts (hunk 4 above)
Comment (chaosmail) [inline src/utils/Options.ts]: it should be `default`, but the keyword is already reserved. i will call it `_default`. but this is a very good point

File: src/utils/SeriesOptions.ts (hunk 5 above)
Comment (chaosmail) [inline src/utils/SeriesOptions.ts]: I was just thinking, why we need to define a main type - we could just add a class for every series type

File: src/utils/Options.ts (hunk 4 above)
Comment (lorem--ipsum) [inline src/utils/Options.ts]: Or `defaultValue` ?

File: src/utils/SeriesOptions.ts (hunk 6 above)
Comment (chaosmail) [inline src/utils/SeriesOptions.ts]: We will never reach this line, so throwing an error is also not the best option I think. We cannot even test this. Hmm that's tricky

File: src/utils/SeriesOptions.ts (hunk 5 above)
Comment (lorem--ipsum) [inline src/utils/SeriesOptions.ts]: Hmmmmm let's hope it's not gonna create conflicts in the CSS...

File: src/utils/SeriesOptions.ts (hunk 6 above)
Comment (lorem--ipsum) [inline src/utils/SeriesOptions.ts]: Then let's remove the last line, and the last `if` ?

File: src/utils/SeriesOptions.ts (hunk 7 above)
Comment (chaosmail) [inline src/utils/SeriesOptions.ts]: it's an ENUM, so you can theoretically also plug values inside. If the values don't match the function will return false anyway. Not sure if this is really necessary here

File: src/utils/Options.ts (hunk 4 above)
Comment (chaosmail) [inline src/utils/Options.ts]: great

File: src/utils/SeriesOptions.ts (hunk 5 above)
Comment (chaosmail) [inline src/utils/SeriesOptions.ts]: I will figure out how is it used, and we keep it open for later ;)

File: src/utils/SeriesOptions.ts (hunk 6 above)
Comment (chaosmail) [inline src/utils/SeriesOptions.ts]: yeah, that's a good idea

File: test/spec/utils/Options.spec.ts (hunk 8 above)
Comment (chaosmail) [inline test/spec/utils/Options.spec.ts]: My idea was to use the 2 variables `testing` and `expected` for every test (except when we test for functions to throw errors). In my opinion the tests get more readable, especially when you have loads and loads of tests.

Comment (chaosmail) [inline test/spec/utils/Options.spec.ts]: I think we can transform all statements into `.equal()` and `.eql()` to improve readability. What do you think?

Comment (lorem--ipsum) [inline test/spec/utils/Options.spec.ts]: Hmmm I don't know if adopting conventions like this one isn't gonna be more constraining than anything else down the road... The type of comparison (`equal`, `eql`, etc.) is often tightly correlated to the tested and expected values, hence I see a real point in not separating them in the code... Also, if there are tools to make the tests more explicit, why not use them if that's only for convention's sake ?

Comment (chaosmail) [inline test/spec/utils/Options.spec.ts]: Yeah it is a constraint, a big one. But with enforcing this structure we improve readability. The problem is also, that there are multiple ways to check the test results, like `euqal`, 'is.a', 'to.have', 'with'.

Comment (chaosmail) [inline test/spec/utils/Options.spec.ts]: Which are these tools to make tests more explicit?

Comment (lorem--ipsum) [inline test/spec/utils/Options.spec.ts]: Well : http://chaijs.com/api/bdd/

Comment (lorem--ipsum) [inline test/spec/utils/Options.spec.ts]: Also, common code structure is not the only way to achieve readability. Concision and general consistency is, in my opinion. So I'd like to loosen up a little bit the tests structure constraints :-)

Comment (chaosmail) [inline test/spec/utils/Options.spec.ts]: I was expecting a super cool tool now, like tests in YAML or sth :smile: 

Ok, I agree with you; I just have 2 more remarks that came to our mind. 
1) We will have loads of contributors, so the testing guidelines should be clear.
2) If we enforce `equal` and `eql`, you only need to look at the 2 variables to understand the test

Comment (lorem--ipsum) [inline test/spec/utils/Options.spec.ts]: Yeah, I get your point, but honestly I can't get my head around having such unflexible constraints for tests ^^
> We will have loads of contributors

Hell yeah !

Comment (chaosmail) [inline test/spec/utils/Options.spec.ts]: Ok, then we agree on a more flexible structure. However, I will try to continue writing tests in this format - and we should decide later on strict testing guidelines (and write them down)

Comment (lorem--ipsum) [inline test/spec/utils/Options.spec.ts]: Alright, we can still make rule more strict if we see that people tend to write crappy tests :-)

**Comment by lorem--ipsum** [comment]
Ready for merge ?

**Comment by chaosmail** [comment]
sure :accept:

**Comment by lorem--ipsum** [comment]
tumblr_ly2ctlp6xh1qzlfumo1_500

**Pull request merged**